lib/Sema/CSGen.cpp

```diff
@@ -4779,15 +4778,15 @@ ConstraintSystem::applyPropertyWrapperToParameter(
       setType(param->getPropertyWrapperProjectionVar(), projectionType);
     }
 
-    initKind = PropertyWrapperInitKind::ProjectedValue;
-  } else {
+    appliedPropertyWrappers[anchor].push_back({ wrapperType, PropertyWrapperInitKind::ProjectedValue });
+  } else if (param->hasExternalPropertyWrapper()) {
```

Comment:
Was this getting called without an external property wrapper to apply from `resolveClosure`? I think that call should perhaps be in an `else` case here:

```cpp
      if (!paramDecl->getName().hasDollarPrefix()) {
        if (generateWrappedPropertyTypeConstraints(paramDecl, backingType,
                                                   param.getParameterType()))
          return false;
      }

      auto result = applyPropertyWrapperToParameter(backingType, param.getParameterType(),
                                                    paramDecl, paramDecl->getName(),
                                                    ConstraintKind::Equal,
                                                    getConstraintLocator(closure));
      if (result.isFailure())
        return false;
```

because `generateWrappedPropertyTypeConstraints` should be sufficient in the case where a property wrapper attribute was written on the closure parameter.

I see another call-site in `unwrapPropertyWrapperParameterTypes`, and I think we might want to adjust the condition right above it to check for `hasExternalProeprtyWrapper()` here:

```cpp
    auto *paramDecl = paramList->get(i);
    if (!paramDecl->hasAttachedPropertyWrapper() && !argLabel.hasDollarPrefix()) {
      adjustedParamTypes.push_back(paramTypes[i]);
      continue;
    }
```

Otherwise, we'll generate extraneous type variables.

The purpose of `applyPropertyWrapperToParameter` is to generate constraints for the inputs, so I think it should fail if there isn't an appropriate property wrapper to apply. Instead of silently returning `getTypeMatchSuccess()`, can you please add back an `else` block that returns `getTypeMatchFailure()`?

Reply:
> Was this getting called without an external property wrapper to apply from `resolveClosure`?

No, in my case it was never called from `resolveClosure`. My knowledge of this codebase is pretty limited at this moment. But, as I get it, `resolveClosure` will be invoked only in case when we are constructing closure from the closure expression `closure_expr`, i.e. 

```swift
let _: (Int) -> Void = { (@Wrapper value) in
  // NoOp
}
```

My case is slightly different because I am constructing closure form dot syntax expression `unresolved_dot_expr`. I have tried to apply your suggestion for `resolveClosure` but it causes older tests to fail.

> I see another call-site in unwrapPropertyWrapperParameterTypes, and I think we might want to adjust the condition right above it to check for `hasExternalProeprtyWrapper()` here:

I think this one is redundant because of `hasExternalPropertyWrapper()` implementation. It will return false if `hasAttachedPropertyWrapper()` returns false. And it is the first part of current if statement. If there are no wrappers, then there will be no external wrapper.

> Instead of silently returning `getTypeMatchSuccess()`, can you please add back an else block that returns `getTypeMatchFailure()`?

Added, tests look good. Thank you for the explanation!